File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
from collections import deque

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        # 기존 문자에서 이미 존재하는 문자가 있다면, 해당 문자까지 잘라내고 더한다.
        # abc 에서 'b' 추가시 -> ab를 잘라내고 b를 추가 -> 'cb'
        # 문자의 set을 만들고, 문자 삭제/추가 할때마다 set의 값도 업데이트 해준다.
        char_set = set()
        queue = deque()
        max_len = 0
        
        for c in list(s):
            if c not in char_set:
                queue.append(c)
                char_set.add(c)
            else:
                while queue[0] != c:
                    char_set.remove(queue.popleft())
                
                queue.popleft()
                queue.append(c)
            
            max_len = max(max_len, len(queue))

        return max_len

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
/*
슬라이딩 윈도우를 어떻게 활용해야 할까?

s = abcabcbb

- n...1 for문을 순회하면서 0부터 i까지의 문자열이 모두 일치하지 않는다면 i 리턴?

*/

class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> set = new HashSet<>();
        Queue<Character> queue = new ArrayDeque<>();
        int maxLength = 0;

        for (char c : s.toCharArray()) {
            if (!set.contains(c)) {
                queue.offer(c);
                set.add(c);
            } else {
                while (queue.peek() != c) {
                    set.remove(queue.poll());
                }

                queue.poll();
                queue.offer(c);
                set.add(c);
            }

            maxLength = Math.max(maxLength, queue.size());
        }

        return maxLength;
    }
}